<a href="https://colab.research.google.com/github/saiprashanthig/Math_Techniques_in_Data_Science/blob/Homework/702687444_Sai_Prashanthi_HW1_F23.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

*   Student name: **[Sai Prashanthi]**
*   UD ID: **[702687444]**
*   Collaborator(s): **[None]**

# MATH637-011 Homework 1 (10% total grade)

## **Deadline: 09/29/2023 11:59 PM**

In this homework, we will train and test simple classification algorithms on a simulated 2d dataset.

## 1. Generate dataset (1% total grade)

Write Python code to generate a classification dataset that contains 10000 examples.

In this dataset, each 2-dimensional input $\textbf{x} = (x_1, x_2)$ is drawn uniformly random from the set $[0, 1] \times [0, 1]$ and the corresponding output is

$y = sign(x_2 - x_1 + \epsilon),$

where $\epsilon$ is Gaussian noise with zero mean and standard deviation 0.1, and
  
  $sign(z) =   \begin{cases}
    1, & \text{for } z>0 \\
    -1, & \text{for } z<0 \\
    0 & \text{for } x=0
  \end{cases}$

In [ ]:
import numpy as np
np.random.seed(7)

In [ ]:
s = np.random.normal(0, 0.1, 10000) #epsilon where we are drawing 10000 samples with mean 0 and std 0.1

In [ ]:
X = np.random.rand(10000, 2)  # Generate samples from a uniform distribution in the range [0, 1]

In [ ]:
X

array([[0.0587769 , 0.29601892],
       [0.77052428, 0.68459312],
       [0.81826552, 0.6124866 ],
       ...,
       [0.59873653, 0.5623609 ],
       [0.95588044, 0.93097251],
       [0.37465482, 0.71987591]])

In [ ]:
z=X[:,1] - X[:,0] + s
Y=np.sign(z)

In [ ]:
Y

array([ 1., -1., -1., ...,  1.,  1.,  1.])

## 2. Split dataset into train/test sets (1% total grade)

Write code to randomly split your dataset above into a train set and a test set. Your train set should contain 7000 examples and your test set should contain 3000 examples.

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state=42) #splitting the dataset

In [ ]:
print("Shape of X_train",X_train.shape)
print("Shape of Y_train",Y_train.shape)
print("Shape of X_test",X_test.shape)
print("Shape of Y_test",Y_test.shape)

Shape of X_train (7000, 2)
Shape of Y_train (7000,)
Shape of X_test (3000, 2)
Shape of Y_test (3000,)


## 3. Define and train a K-NN model (1% total grade)

Write code to define and train a K-NN classification model with K=5 and the neighbors are weighted by the inverse of their distance.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
clock= KNeighborsClassifier( n_neighbors=5,weights='distance')

In [ ]:
clock.fit(X_train,Y_train)

KNeighborsClassifier(weights='distance')

## 4. Make prediction with your K-NN model (1% total grade)

Write code to predict the outputs of your K-NN model on the test set.

In [ ]:
Y_pred=clock.predict(X_test)

## 5. Compute the accuracy (1% total grade)

Write code to compute and print out the accuracy of your K-NN model on the test set

In [ ]:
from sklearn.metrics import accuracy_score
acc_knn= accuracy_score(Y_test,Y_pred)

In [ ]:
print("KNN accuracy",round(acc_knn,4))

KNN accuracy 0.9083


## 6. Repeat the experiments with a logistic regression model (2% total grade)


Repeat Steps 3-5 for a logistic regression using an $l_1$ penalty with regularizing constant C=5.0.  You may need to choose a non-default optimization solver for the $l_1$ penalty.

In [ ]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(penalty='l1',C=5.0,solver='saga')

In [ ]:
clf.fit(X_train,Y_train)
Y_pred_log=clf.predict(X_test)

In [ ]:
acc_log=accuracy_score(Y_test, Y_pred_log)
print("Logistic regression accuracy", round(acc_log,4))

Logistic regression accuracy 0.9207


## 8. Repeat the experiments with a SCM Classifier

Repeat Steps 3-5 for a SVM Classifier with polynomial kernel (with degree 2) and regularization parameter C=1  (2% total grade)

In [ ]:
from sklearn.svm import SVC
svm = SVC( kernel='poly', degree=2,  C=1.0)

In [ ]:
svm.fit(X_train,Y_train)

SVC(degree=2, kernel='poly')

In [ ]:
Y_pred_svc=svm.predict(X_test)
acc_scm= accuracy_score(Y_test, Y_pred_svc)
print("SCM classifier", round(acc_scm,4))

SCM classifier 0.9217


## 10. Compare the three classifiers (1% total grade)

According to the above results, which model is better on your dataset?

**Your answer**: SVM classifier has an accuracy of 92.17 %, Logistic regression has an accuracy of 92.07%, KNN has an accuracy of 90.83%. According to the above results, SVM classifier performs better on our dataset as it has highest accuracy of 92.17% followed by Logistic regression which has almost similar accuracy of 92.07% followed by KNN.